File: Review-04-Unit-Testing/Example-4/src/test/java/TestBoard.java
import org.junit.jupiter.api.TestMethodOrder;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;

import static org.junit.jupiter.api.Assertions.*;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import org.hamcrest.core.IsNull;

/**
 * 1 - Does this piece of code perform the operations
 *     it was designed to perform?
 *
 * 2 - Does this piece of code do something it was not
 *     designed to perform?
 *
 * 1 Test per mutator
 */
@TestMethodOrder(MethodOrderer.MethodName.class)
public class TestBoard
{
    @Test
    public void testBoardButPoorly()
    {
        Board board = new Board();

        String expectedInitial = "1|2|3\n4|5|6\n7|8|9\n";
        String expectedFinal   = "X|X|X\nX|X|X\nX|X|X\n";

        assertThat(board.toString(), equalTo(expectedInitial));
        assertFalse(board.isFull());

        for (int i = 0 ; i < 9; i++) {
            board.setCell((i + 1), 'X');
        }

        assertThat(board.toString(), equalTo(expectedFinal));
        assertTrue(board.isFull());
    }

    // Where are the rest of the tests? This might
    // be worth an F... if I am lucky.

}

File: Review-04-Unit-Testing/Example-1-CPP/source/TestBoard.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstdlib>
#include <vector>
#include <sstream>
#include <algorithm>
#include <functional>

#include "Board.h"

#include "utilities.h"
#include "bodgeUnitTest.h"

//-----------------------------------------------------------------------------
// Unit Tests - Support Data
//-----------------------------------------------------------------------------

//------------------------------------------------------------------------------
// Unit Tests - Test Functions
//------------------------------------------------------------------------------

bool testBoardButPoorly()
{
    Board board;

    std::string expectedInitial = "1|2|3\n4|5|6\n7|8|9\n";
    std::string expectedFinal   = "X|X|X\nX|X|X\nX|X|X\n";

    bodgeAssert(toStr(board) == expectedInitial);
    bodgeAssert(!board.isFull());

    for (int i = 0 ; i < 9; i++) {
        board.setCell((i + 1), 'X');
    }

    bodgeAssert(toStr(board) == expectedFinal);
    bodgeAssert(board.isFull());

    return true;
}

// Where are the rest of the tests? This might
// be worth an F... if I am lucky.

//-----------------------------------------------------------------------------
int main(int argc, char** argv)
{
    UnitTestPair tests[] = {
        {testBoardButPoorly, "testBoardButPoorly"}
    };

    runTests(tests);

    return 0;
}
